st2api/st2api/controllers/exp/inquiries.py

```diff
@@ -0,0 +1,295 @@
+            Inquiries so we don't return a bunch of extra crap. The idea is to
+            provide data in response to requests that make it look like Inquiry
+            is its own data model.
+        """
+
+        # The "put" and "get" commands will have an actual ActionExecutionAPI object, so
+        # if that's what's being passed in, let's convert it to a dict first.
+        # ("get_all" provides a dict natively)
+        if isinstance(raw_inquiry, ActionExecutionAPI):
+            raw_inquiry = raw_inquiry.__json__()
+
+        desired_fields = ["tag", "ttl", "users", "roles", "schema"]
+
+        inquiry = {
+            "id": raw_inquiry["id"]
+        }
+        for field in desired_fields:
+            inquiry[field] = raw_inquiry["result"][field]
+
+        return inquiry
+
+    def _mark_inquiry_complete(self, inquiry_id, result):
+        """Mark Inquiry as completed
+
+        This function updates the local LiveAction and Execution with a successful
+        status as well as call the "post_run" function for the Inquirer runner so that
+        the appropriate callback function is executed
+
+        :param inquiry: The Inquiry for which the response is given
+        :param requester_user: The user providing the response
+
+        :rtype: bool - True if requester_user is able to respond. False if not.
+        """
+
+        # Update inquiry's execution result with a successful status and the validated response
+        liveaction_db = action_utils.update_liveaction_status(
+            status=action_constants.LIVEACTION_STATUS_SUCCEEDED,
+            runner_info=system_info.get_process_info(),
+            result=result,
+            liveaction_id=inquiry_id)
+        executions.update_execution(liveaction_db)
+
+        # Call Inquiry runner's post_run to trigger callback to workflow
+        runner_container = get_runner_container()
+        action_db = get_action_by_ref(liveaction_db.action)
+        runnertype_db = get_runnertype_by_name(action_db.runner_type['name'])
+        runner = runner_container._get_runner(runnertype_db, action_db, liveaction_db)
+        runner.post_run(status=action_constants.LIVEACTION_STATUS_SUCCEEDED, result=result)
+
+        return liveaction_db
+
+    def _can_respond(self, inquiry, requester_user):
+        """Determine if requester_user is permitted to respond based on parameters
+
+        This determines if the requesting user has permission to respond to THIS inquiry.
+        Note this is NOT RBAC, and you should still protect the API endpoint with RBAC
+        where appropriate.
+
+        :param inquiry: The Inquiry for which the response is given
+        :param requester_user: The user providing the response
+
+        :rtype: bool - True if requester_user is able to respond. False if not.
+        """
+
+        # Deny by default
+        roles_passed = False
+        users_passed = False
+
+        # Determine role-level permissions
+        roles = inquiry.get('roles')
```

Comment:
I prefer to do `roles = inquiry.get('roles', [])` and then just `for role in roles` to avoid if and additional level of nesting, but that's just a personal preference and current approach works as well :)

Reply:
The catch here was that breaking out of the loop meant you were permitted, so it was difficult to JUST rely on the loop for this. However, I think I carried the spirit of what you're saying into 39a17fa5d - I'm fairly confident the logic is the same but wouldn't mind a sanity check